### PR TITLE
Next/Previous card gestures in Previewer

### LINKED ISSUE #12390: Lack of gestures in preview window
I'm always frustrated when I preview a deck and i cant go to the next card with a single hand because the go to the next card is a small button on the bottom right corner

I would like to be able to tap or swipe right to go to the next card

I have tried to make the buttons bigger, but holding the phone with the right hand is not helping that. The show answer button is bigger and i can almost press it comfortably press it while holding my phone with my right hand, but it wont go to the next card

The reason im using this is because Im using the preview window to show all cards of a deck to review something learned with [LPCG (Lyrics/Poetry Cloze Generator)](https://ankiweb.net/shared/info/2084557901) after learning every individual card using regular anki

### PR DESCRIPTION
**Pull request by alyoshenka**
## Purpose / Description
This fix introduces gesture support for viewing the next and previous card in `Previewer`.

## Fixes
Fixes #12390 

## Approach
This adds the relevant viewer commands and checks for them in `Previewer` when a gesture is detected.

From what I could see, there are no other commands that have confusingly similar meanings. To the best of my knowledge, these new commands are set up like the others, however I may have missed something.

As per the conversation in the original issue, there are more commands that could be added to the `Previewer`, such as flipping the card. This pr just fixes the main issue, but I am looking into others to add as well. 

## How Has This Been Tested?
1. Go to Settings > Controls > Set 'Preview Next/Previous Card' to the desired gesture
2. Go to Previewer
  1. Card Browser > Options Menu > Preview
- [x] The selected gestures move to next/previous card 
  - [x] 'Previous' gesture does nothing when at the beginning of the deck
  - [x] 'Next' gesture does nothing when at the end of the deck

## Learning (optional, can help others)
I couldn't find documentation detailing how to add new gestures. It seemed pretty straightforward, but it would be worthwhile to add a documentation page for it. 

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] ~UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)~
- [x] ~UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)~

### CODE REVIEW

**Comment by BrayanDSO** [comment]
Hi. I think that the user wants to re-use the gestures they use on the reviewer instead of creating duplicates exclusive to the previewer.

So I believe that the goal here is to make some gestures like `Show answer` and `Answer button X` to work on the previewer as well. 

As you are the issue creator, do you have any ideas @voidplayer ?

And oh, tests are failing

**Comment by alyoshenka** [comment]
What actions would showing the next and previous card map to? I couldn't find any that seemed appropriate.

File: AnkiDroid/src/main/java/com/ichi2/anki/Previewer.kt [hunk 1]
```diff
@@ -234,6 +243,17 @@ class Previewer : AbstractFlashcardViewer() {
         mNoteChanged = true
     }
 
+    private fun tryChangePreviewedCard(nextCard: Boolean): Boolean {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/Previewer.kt:246]: This method name and parameter are a bit confusing, although I know that the logic is similar to the existing `changePreviewedCard` method

Perhaps having separate methods named `showNextCard` and `showPreviousCard` (or similar names) without a parameter would be better

File: AnkiDroid/src/main/res/values/11-arrays.xml [hunk 2]
```diff
@@ -99,6 +99,8 @@
     <string name="gesture_toggle_whiteboard">Toggle Whiteboard</string>
     <string name="gesture_show_hint">Show Hint</string>
     <string name="gesture_show_all_hints">Show All Hints</string>
+    <string name="gesture_preview_next_card">Preview Next Card</string>
+    <string name="gesture_preview_previous_card">Preview Previous Card</string>
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/res/values/11-arrays.xml:103]: I know that the other gestures have all words capitalized, but they shouldn't be, so I'd rather if these were `Preview next card` and `Preview previous card`

**Review by BrayanDSO — COMMENTED** [review]
In order to avoid people getting confused, it would be best to create a `PreferenceCategory` called `Previews` so the user know that these are specific to the previewer.

If this is created, maybe the strings could be simplified to `Next card` and `Previous card`.

---
For other issues:
1. Reuse the existing gestures as well (Answer button 1/2/3/4 and show answer)
2. Make the previewer answer bar be resizable as well

File: AnkiDroid/src/main/java/com/ichi2/anki/Previewer.kt (hunk 1 above)
Comment (alyoshenka) [reply AnkiDroid/src/main/java/com/ichi2/anki/Previewer.kt]: Updated in most recent commit.

File: AnkiDroid/src/main/res/values/11-arrays.xml (hunk 2 above)
Comment (alyoshenka) [reply AnkiDroid/src/main/res/values/11-arrays.xml]: Updated in most recent commit.

**Comment by alyoshenka** [comment]
> Yes, ideally same gestures work without having to define them
> 
> but `go to next/previous card without showing answer` is an action exclusive of the preview window and have no gestures

For this issue specifically, I'm working on showing next/prev card. If I get around to supporting other gestures, I'll make a separate PR.

**Comment by alyoshenka** [comment]
After looking more at `ViewerCommand`, it does seem like this class is intended to be used only with `Reviewer`: all of the MapableBindings are connected to it. I feel like this means a larger redesign is necessary, but I'm not sure what that entails.

**Comment by alyoshenka** [comment]
My solution that seems to work is as follows:
- keep all gesture actions in `ViewerCommand.values()`
- add specific gestures for each `AbstractFlashcardViewer` (just `Reviewer` and `Previewer` in this case) to the companion object
  - so you would access previewer gestures with `ViewerCommand.previewCommands`
  - This has the downside of needing to repeat most commands into `ViewerCommand.reviewerCommands`, but the upside of making an easy place to define which commands go to which viewer
- add a new `PreferenceCategory` to `ControlsSettingsFragment` and populate it with the Previewer commands
This comes with the downside that there would need to be a new translated string resource: "Previewer commands" (and potentially also changing "Command mapping" to "Reviewer commands'). However, this would be necessary anyway with the necessity of making the Previewer command category.

I'm working on getting this change ready to push, but it will take some time to get the string resources figured out. That's my solution, at least.

**Comment by alyoshenka** [comment]
There is also an issue where pressing a key in `Previewer` causes the screen to darken. This is an issue in `main` as well.

**Comment by alyoshenka** [comment]
This PR still needs translations. It is my understanding that now that I've added a string resource, it will go out for translation, and I don't have to do anything further.

File: AnkiDroid/src/main/java/com/ichi2/anki/preferences/ControlsSettingsFragment.kt [hunk 3]
```diff
@@ -28,13 +28,17 @@ class ControlsSettingsFragment : SettingsFragment() {
...
+        // Previewer-specific commands
+        val commandMappingCategoryPreviewer = requirePreference<PreferenceCategory>(R.string.controls_command_mapping_cat_key_previewer)
+        addAllControlPreferencesToCategory(commandMappingCategoryReviewer, ViewerCommand.reviewerCommands)
+        addAllControlPreferencesToCategory(commandMappingCategoryPreviewer, ViewerCommand.previewerCommands)
     }
 
     /** Attaches all possible [ControlPreference] elements to a given [PreferenceCategory] */
-    fun addAllControlPreferencesToCategory(category: PreferenceCategory) {
-        for (command in ViewerCommand.values()) {
+    fun addAllControlPreferencesToCategory(category: PreferenceCategory, commands: List<ViewerCommand>) {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/preferences/ControlsSettingsFragment.kt:40]: Renaming the method to `addCommandsToCategory` (or something similar) would make more sense now.

And rewriting the javadoc would be appropriate too.
```kotlin
    /** 
       * Creates dinamically [ControlPreference] objects for each of the [commands] given,
       * and add them to [category]
       */
```

File: AnkiDroid/src/main/java/com/ichi2/anki/preferences/Preferences.kt [hunk 4]
```diff
@@ -151,6 +151,7 @@ class Preferences : AnkiActivity(), SearchPreferenceResultListener {
                 .withResId(R.xml.preferences_controls)
                 .addBreadcrumb(getString(R.string.pref_cat_controls))
                 .addBreadcrumb(getString(R.string.controls_main_category))
+                .addBreadcrumb(getString(R.string.controls_previewer))
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/preferences/Preferences.kt:154]: The preferences search bar indexes preferences from the XMLs. As we build controlPreferences dinamically, they aren't on the XMLs and therefore aren't indexed with the others, so we have to index them manually like on this block. 

Note that the breadcrumbs should be different between reviewer and previewer methods, as they are on different categories, so basically you need to have a different loop for each of them. So I suggest something like

```kotlin
for (command in ViewerCommand.reviewerCommands) {
            searchConfig.indexItem()
                .withTitle(getString(command.resourceId))
                .withKey(command.preferenceKey)
                .withResId(R.xml.preferences_controls)
                .addBreadcrumb(getString(R.string.pref_cat_controls))
                .addBreadcrumb(getString(R.string.controls_main_category))

for (command in ViewerCommand.previewerCommands) {
            searchConfig.indexItem()
                .withTitle(getString(command.resourceId))
                .withKey(command.preferenceKey)
                .withResId(R.xml.preferences_controls)
                .addBreadcrumb(getString(R.string.pref_cat_controls))
                .addBreadcrumb(getString(R.string.controls_previewer))
```

File: AnkiDroid/src/main/res/values/preferences.xml [hunk 5]
```diff
@@ -77,6 +77,7 @@
     <string name="nav_drawer_gesture_key">gestureFullScreenNavigationDrawer</string>
     <string name="pref_swipe_sensitivity_key">swipeSensitivity</string>
     <string name="controls_command_mapping_cat_key">key_map_category</string>
+    <string name="controls_command_mapping_cat_key_previewer">key_map_category_previewer</string>
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/res/values/preferences.xml:80]: Have `_key` at the end of the key

File: AnkiDroid/src/main/res/values/10-preferences.xml [hunk 6]
```diff
@@ -235,6 +235,7 @@
     <string name="keyboard">Keyboard</string>
     <string name="bluetooth">Bluetooth</string>
     <string name="controls_main_category" maxLength="41">Command mapping</string>
+    <string name="controls_previewer" maxLength="41">Previewer commands</string>
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/res/values/10-preferences.xml:238]: `controls_previewer_cat` or `controls_previewer_category`

**Review by BrayanDSO — CHANGES_REQUESTED** [review]
Overall looks pretty good. Great job!

Of my remarks, probably only the search bar breadcrumbs one is important, rest of them are nits.

And I agree with you that `Command mapping` should be renamed to `Reviewer commands` (remember to change the key too when you do it)

**Comment by alyoshenka** [comment]
Sweet! Didn't expect everything to be 100% the first time, glad to hear this approach is acceptable. I'll get those changes in as soon as I can.

**Comment by alyoshenka** [comment]
Not sure what's going on with ubuntu-latest tests, I ran them locally with no errors.

**Comment by dae** [comment]
Network flake; restarted.

File: AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/ViewerCommand.kt [hunk 7]
```diff
@@ -78,6 +78,52 @@ enum class ViewerCommand(val resourceId: Int) {
             get() = Arrays.stream(values())
                 .flatMap { x: ViewerCommand -> x.defaultValue.stream() }
                 .collect(Collectors.toList())
+
+        val reviewerCommands: List<ViewerCommand>
```
[the remark this answers is not in the document]
Comment (alyoshenka) [reply AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/ViewerCommand.kt:82]: Covered in most recent commit

File: AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/ViewerCommand.kt [hunk 8]
```diff
@@ -166,6 +212,8 @@ enum class ViewerCommand(val resourceId: Int) {
                 SHOW_HINT -> from(keyCode(KeyEvent.KEYCODE_H, CardSide.BOTH))
                 SHOW_ALL_HINTS -> from(keyCode(KeyEvent.KEYCODE_G, CardSide.BOTH))
                 ADD_NOTE -> from(keyCode(KeyEvent.KEYCODE_A, CardSide.BOTH))
+                SHOW_NEXT_CARD -> from(keyCode(KeyEvent.KEYCODE_P, CardSide.BOTH))
+                SHOW_PREV_CARD -> from(keyCode(KeyEvent.KEYCODE_O, CardSide.BOTH))
```
[the remark this answers is not in the document]
Comment (alyoshenka) [reply AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/ViewerCommand.kt]: Covered in most recent commit

File: AnkiDroid/src/main/java/com/ichi2/anki/preferences/Preferences.kt (hunk 4 above)
Comment (alyoshenka) [reply AnkiDroid/src/main/java/com/ichi2/anki/preferences/Preferences.kt]: Covered in most recent commit

File: AnkiDroid/src/main/res/values/10-preferences.xml (hunk 6 above)
Comment (alyoshenka) [reply AnkiDroid/src/main/res/values/10-preferences.xml]: Covered in most recent commit

**Review by BrayanDSO — CHANGES_REQUESTED** [review]

**Review by david-allison — CHANGES_REQUESTED** [review]
This should be using the `MappableBinding.Screen` class to serialise properties designed for a new screen

**Comment by david-allison** [comment]
I'll take this one to completion in 2.17 if there's no reply. Good feature to have.

**Pull request closed without merging**